onnxruntime/python/tools/quantization/shape_inference.py

```diff
@@ -0,0 +1,261 @@
+    """Shape inference and model optimization, in preparation for quantization.
+
+    Args:
+        input_model_path: Path to the input model file")
+        output_model_path: Path to the output model file
+        skip_optimization: Skip model optimization step if true. This may result in ONNX shape
+            inference failure for some models.
+        skip_onnx_shape: Skip ONNX shape inference. Symbolic shape inference is most effective
+            with transformer based models. Skipping all shape inferences may
+            reduce the effectiveness of quantization, as a tensor with unknown
+            shape can not be quantized.
+        skip_symbolic_shape: Skip symbolic shape inference. Symbolic shape inference is most
+            effective with transformer based models. Skipping all shape
+            inferences may reduce the effectiveness of quantization, as a tensor
+            with unknown shape can not be quantized.
+        auto_merge: For symbolic shape inference, automatically merge symbolic dims when
+            confliction happens.
+        int_max: For symbolic shape inference, specifiy the maximum value for integer to be
+            treated as boundless for ops like slice
+        guess_output_rank: Guess output rank to be the same as input 0 for unknown ops
+        verbose: Logs detailed info of inference, 0: turn off, 1: warnings, 3: detailed
+        save_as_external_data: Saving an ONNX model to external data
+        all_tensors_to_one_file: Saving all the external data to one file
+        external_data_location: The file location to save the external file
+        external_data_size_threshold: The size threshold for external data
+    """
+    with tempfile.TemporaryDirectory(prefix="pre.quant.") as quant_tmp_dir:
+        temp_path = Path(quant_tmp_dir)
+        model = None
+
+        if not skip_symbolic_shape:
+            logger.info("Performing symbolic shape inference...")
+            model = SymbolicShapeInference.infer_shapes(
+                onnx.load(input_model_path),
+                int_max,
+                auto_merge,
+                guess_output_rank,
+                verbose,
+            )
+
+        if not skip_optimization:
+            # Use ORT optimizers (native code) to optimize model
+            if not skip_symbolic_shape:
+                # Need to save the inferenced model to file so as to run the optimizer
+                input_model_path = str(temp_path / "symbolic_shape_inferred.onnx")
+                onnx.save(model, input_model_path)
+                model = None
+
+            opt_model_path = str(temp_path / "optimized.onnx")
+            sess_option = onnxruntime.SessionOptions()
+            sess_option.optimized_model_filepath = opt_model_path
+            sess_option.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC
+            _ = onnxruntime.InferenceSession(input_model_path, sess_option, providers=["CPUExecutionProvider"])
```

Comment:
if model is larger than 2gb, this will fail

Reply:
Is it because ORT native code can not output model with external data and the 2GB limit of protobuf? In that case we can't really do anything. And that's the reason we have a "skip optimization" option here.

Reply:
It's hard to predict the fail before calling it. Since optimization would change the graph, the resulting model size would change, right? So the only option we have is to write something in docs?